CIP3/CIP3.md

```diff
@@ -0,0 +1,246 @@
+
+### Master Key Generation
+
+The master key generation is the mean by which on turns an initial entropy into a secure cryptographic key.
+
+More specifically, the generation is a function from an initial seed to an extended private key (abbrev. XPrv) composed of:
+
+- 64 bytes: an extended Ed25519 secret key composed of:
+  - 32 bytes: Ed25519 curve scalar from which few bits have been tweaked according to ED25519-BIP32
+  - 32 bytes: Ed25519 binary blob used as IV for signing
+- 32 bytes: chain code for allowing secure child key derivation
+
+#### History
+
+Throughout the years, Cardano has used different styles of master key generation:
+
+|  Name          |  Created by |  Address prefix in Byron |  Is deprecated? |
+|----------------|-------------|--------------------------|-----------------|
+|  Byron         |  Daedalus   |  Ddz                     |  Yes            |
+|  Icarus        |  Yoroi      |  Ae2                     |  No             |
+|  Icarus-Trezor |  Trezor     |  Ae2                     |  No             |
+|  Ledger        |  Ledger     |  Ae2                     |  No             |
+
+# Rationale
+
+This CIP is merely to document the existing standards and not to provide rationales for the various methods used.
+
+However, you can learn more at the following links:
+
+- [Adrestia documentation](https://input-output-hk.github.io/adrestia/docs/key-concepts/hierarchical-deterministic-wallets/)
+- [SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
+- [SLIP-0023](https://github.com/satoshilabs/slips/blob/master/slip-0023.md)
+
+## Reference implementations
+
+We provide a description, reference implementation and test vectors for each algorithm
+
+#### Byron
+
+```js
+function generateMasterKey(seed) {
+    return hashRepeatedly(seed, 1);
+}
+
+function hashRepeatedly(key, i) {
+    (iL, iR) = HMAC
+        ( hash=SHA512
+        , key=key
+        , message="Root Seed Chain " + UTF8NFKD(i)
+        );
+
+    let prv = tweakBits(SHA512(iL));
+
+    if (prv[31] & 0b0010_0000) {
+        return hashRepeatedly(key, i+1);
+    }
+
+    return (prv + iR);
+}
+
+function tweakBits(data) {
+    // * clear the lowest 3 bits
+    // * clear the highest bit
+    // * set the highest 2nd bit
+    data[0]  &= 0b1111_1000;
+    data[31] &= 0b0111_1111;
+    data[31] |= 0b0100_0000;
+
+    return data;
+}
+```
+
+##### Test vectors
+
+```
+mnemonic: install neither high supreme hurdle tissue excite lava census harbor purpose shine
+master key: 309c18090ec559d4c8e377b9b19b84804a4fc430e862ef99d97f1cceccad124d92d7336d2d598affefc05c806fe2ff71cfc0fb78bfdb539178dbfcc567a04e358077d66a909e93f4577b5784b8c1c4ff3c77e706449f1c21793a79fc2abb68a9
+```
```

Comment:
It'd be nice to have multiple test vectors which highlight:

- at least one repetition, i.e. where the 5th bit of the last byte of the private key is originally set. 
- no repetition

Reply:
```
# Requires no iteration
["roast","crime","bounce","convince","core","happy","pitch","safe","brush","exit","basic","among"]
60f6e2b12f4c51ed2a42163935fd95a6c39126e88571fe5ffd0332a4924e5e5e9ceda72e3e526a625ea86d16151957d45747fff0f8fcd00e394b132155dfdfc2918019cda35f1df96dd5a798da4c40a2f382358496e6468e4e276db5ec35235f
```

```
# Requires 4 iterations
["legend","dismiss","verify","kit","faint","hurdle","orange","wine","panther","question","knife","lion"]
c89fe21ec722ee174be77d7f91683dcfd307055b04613f064835bf37c58f6a5f362a4ce30a325527ff66b6fbaa43e57c1bf14edac749be3d75819e7759e9e6c82b264afa7c1fd5b3cd51be3053ccbdb0224f82f7d1c7023a96ce97cb4efca945
```